File: findKthNumber.py
# 不知道是哪个
class Solution1:
    def findKthNumber(self, m: int, n: int, k: int) -> int:
        def count(a):
            out = 0

            for i in range(1, n + 1):
                out += min(a // i, m)

                if a < i:
                    break

            return out

        l = 1
        r = m * n

        while l <= r:
            mid = (l + r) // 2
            tmp = count(mid)

            if tmp < k:
                l = mid + 1
            else:
                r = mid - 1
            print(l, r)

        return l


# sl=Solution()
# m = 2
# n = 3
# k = 6
# print(sl.findKthNumber(m,n,k))


# 字典序的第K小数字
class Solution:
    def findKthNumber(self, n: int, k: int) -> int:
        def count(prefix, n):
            cur = prefix
            nxt = prefix + 1
            cnt = 0

            while cur <= n:
                cnt += min(n + 1, nxt) - cur
                cur *= 10
                nxt *= 10

            return cnt

        out = 1
        cnt = 1

        while cnt < k:
            tmp = count(out, n)

            if cnt + tmp > k:
                out *= 10
                cnt += 1
            else:
                out += 1
                cnt += tmp

        return out


n = 13
k = 2
print(Solution().findKthNumber(n, k))

File: findKthNumber.c
#include"ctools.h"
#include<stdio.h>
#include<stdlib.h>
#include<string.h>

//#define testMod
#ifdef testMod
void test() 
{

}
#endif

#define max(a,b) (((a)>(b))?(a):(b))
#define min(a,b) (((a)<(b))?(a):(b))

#ifndef testMod
// your code here
int findKthNumber(int m, int n, int k){
    int left=1,right=30000*30000;
    while(left<=right) {
        int mid=(left+right)/2;
        int tmp=0;
        for(int i=1;i<=n;i++) {
            tmp+=min(mid/i,m);
            if(mid<=i)
                break;
        }
        if(tmp>=k)
            right=mid-1; 
        else
            left=mid+1;
    }
    return left;
}

#endif

int main()
{
#ifdef testMod
    test();
#endif
#ifndef testMod
    int m = 3, n = 3, k = 5;
    m = 2, n = 3, k = 6;
    printf("%d\n",findKthNumber(m,n,k));
#endif
    return 0;
}
    
